### PR TITLE
Collect more perf data for TreesLatest call - Add remote origin and Content-Encoding info #9399

### PR DESCRIPTION
**Pull request by NicholasCouri**
Fix for #9399 

## Description

There is a push to optimize boot performance, specifically P95 numbers.
Snapshot fetch is one of the biggest contributors and, in order to better understand the improvements that need to happen. we need to collect more data around the scenario: 
- Content-Encoding 
- Additional data to record whether the data is coming from Redis cache on SPO side or not. (responseOrigin == graph or cache)
Now the work item https://onedrive.visualstudio.com/SPIN/_workitems/edit/1309956   has been checked in, we can add this information on our end.  @vladsud   - I could simply upload the entire key value pair to the kusto db but I thought it would be better to add a field for each value key pair. As of now there is only one but it could easily change in the future as we add new keys to the bag).

## PR Checklist

> Use the check-list below to ensure your branch is ready for PR. If the item is not applicable, leave it blank.

-   [ ] I have updated the documentation accordingly.
-   [ ] I have added tests to cover my changes.
-   [ ] All new and existing tests passed.
-   [x] My code follows the code style of this project.
-   [ ] I ran the lint checks which produced no new errors nor warnings for my changes.
-   [x] I have checked to ensure there aren't other open Pull Requests for the same update/change.

## Does this introduce a breaking change?

-   [ ] Yes
-   [X] No

### CODE REVIEW

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts [hunk 1]
```diff
@@ -30,7 +30,9 @@ export function getSPOAndGraphRequestIdsFromResponse(headers: { get: (id: string
     interface LoggingHeader {
         headerName: string;
         logName: string;
+        mapping?: Map<string, string | number | boolean>;
```
Comment (vladsud) [inline packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:33]: I'm not sure what this argument is about - it's not used, and nobody provides it

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts [hunk 2]
```diff
@@ -50,6 +54,31 @@ export function getSPOAndGraphRequestIdsFromResponse(headers: { get: (id: string
...
+        const fluidKeyValuesToLog: LoggingHeader[] = [
+            { headerName: "Origin", logName: "responseOrigin", mapping: mapResponseOrigin },
+        ];
+
+        const keyValueMap = fluidTelemetry.split(",").map((keyValuePair) => keyValuePair.split("="));
+        for (const [key, value] of keyValueMap) {
+            const fluidKV = fluidKeyValuesToLog.find((t) => t.headerName === key.trim());
+            if (fluidKV !== undefined) {
+                const fieldValue = value?.trim() ?? true; // assume true for a key without value.
+                additionalProps[fluidKV.logName] = fluidKV.mapping?.get(fieldValue) ?? fieldValue;
```
Comment (vladsud) [inline packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:77]: I'm lost on this line. :)
dKV.mapping?.get(fieldValue) will give us eiter "cache" or "graph", i.e. pre-computed names.
fieldValue will give as actual values.
I do not get why it's Ok to use them interchangeably here (i.e. occasionally record one or another)

Comment (NicholasCouri) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:77]: Yes, the idea is to use the field value as default unless we have a mapping, like we do for c/cache g/graph. 
Eventually we can ask ODSP folks to add new fields and they do not need to have a mapping. Does this help ?

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts (hunk 1 above)
Comment (NicholasCouri) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:33]: I use it to map the x-fluid-telemetry field below (mapping). I was initially adding a new interface specifically for the x-fluid-telemetry but it can be eventually used on the headersToLog as well. So I simply added the mapping property as optional.

Comment (vladsud) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:33]: Ah, I see it (missed it). I'd say - I do not see much value in adding it here.
You can simply remove this addition and do 

const fluidKeyValuesToLog = {...} 

below without any type
I.e. I'm not even sure why this type is related to your changes, it does not affect how actual first-level properties are passed.

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts (hunk 2 above)
Comment (vladsud) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:77]: I'd say it is very hard to parse. And I'm not sure it's clear to readers when and why certain values are used. I think this is more readable:
```ts
const keyValueMap = fluidTelemetry.split(",").map((keyValuePair) => keyValuePair.split("="));
for (const [key, value] of keyValueMap) {
	if ("Origin" === key.trim()) {
		let value: string;
		switch (value?.trim()) {
			case "c": 
				value = "cache";
				break;
			case "g": 
				value = "graph";
				break;
			default:
				value = true;
		}
		additionalProps["responseOrigin"] = value;
	}
}
```
I'd not try to optimize for the next case like that - it's very unlikely to ever exists, and even if it does, I'd still prefer easier to follow code. And it's actually less code, and way easier to follow the intention.

I'm not sure that "true" is the right type here. I'd rather write "undefined" (as a string) or some other string that better represents intention. Maybe not recording it at all (i.e. using actual undefined) is Ok, given that we did not learn anything, so it's the case as Origin was not provided at all.

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts [hunk 3]
```diff
@@ -50,6 +53,32 @@ export function getSPOAndGraphRequestIdsFromResponse(headers: { get: (id: string
...
+                let fieldValue: string;
+                switch (value?.trim()) {
+                    case "c":
+                        fieldValue = "cache";
+                    break;
+                    case "g":
+                        fieldValue = "graph";
+                    break;
+                    default:
+                        fieldValue = "undefined";
```
Comment (vladsud) [inline packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:74]: Let's use "none" or "n/a", I think it reads better by humans. :)

**Review by vladsud — APPROVED** [review]

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts (hunk 3 above)
Comment (vladsud) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts]: I'm fine with that.
